### ISSUE TITLE
Add support for publishing mentions

### ISSUE THREAD

**Issue by gRegorLove** [snarfed/bridgy-fed#493]
I'd like to be able to publish a note to start a conversation with someone (vs replying to a post in a thread). Also handy for tagging someone in a post in an existing thread.

I tried a couple ways a while back but they didn't show up as notifications for the person I mentioned:

https://gregorlove.com/2022/11/hello-from-my-website/ - linked the @-@
https://gregorlove.com/2022/11/hello-again-from-my-website/ - no @-@ link

**Comment by snarfed** [comment]
Yes! Great point, and BF already does the other direction, AP => wm (background in eg #270), so it should absolutely do this too.

**Comment by snarfed** [comment]
More motivation for this: the emerging standard way to do groups on the fediverse is evidently accounts that automatically repost you when you mention them, eg https://venera.social/profile/activitypubblueskybridge

**Comment by snarfed** [comment]
I've been trying to get this to work with a Friendica group, which works by reposting your posts that mention it, but no luck so far. Here are all the different AS2 `Mention` `tag`s I've tried:

```json
    "tag": [{
      "type": "Mention",
      "id": "https://venera.social/profile/activitypubblueskybridge",
      "name": "AP-AT-Bridge Group"
    }]
```

```json
    "tag": [{
      "type": "Mention",
      "id": "https://venera.social/profile/activitypubblueskybridge",
      "url": "https://venera.social/profile/activitypubblueskybridge",
      "name": "@activitypubblueskybridge@venera.social"
    }]
```

```json
    "tag": [{
      "type": "Mention",
      "id": "https://venera.social/profile/activitypubblueskybridge",
      "url": "https://venera.social/profile/activitypubblueskybridge",
      "href": "https://venera.social/profile/activitypubblueskybridge",
      "name": "@activitypubblueskybridge@venera.social"
    }]
```

...all with `cc <a href=\"https://venera.social/profile/activitypubblueskybridge\">@activitypubblueskybridge@venera.social</a>` in `content` and with this audience targeting:

```json
  "to": [
    "https://www.w3.org/ns/activitystreams#Public",
    "https://venera.social/profile/activitypubblueskybridge"
  ]
```

🤷 🤷 🤷

**Comment by snarfed** [comment]
Mastodon docs for mentions: https://docs.joinmastodon.org/spec/activitypub/#Mention

I guess I could try some of the above ^ examples with `cc` instead of `to`.

**Comment by snarfed** [comment]
Tried again with `cc`, delivered this to https://indieweb.social/users/snarfed/inbox just now, it mentions and cc's https://indieweb.social/users/snarfed , but still no luck, that account didn't get a notification.

Next step is to hope @manton can help, he evidently has outgoing fediverse mentions working in micro.blog!

```json
{
  "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-12_snarfed-foo-bar-mas-to#bridgy-fed-update-2023-10-12T23:44:49.320885+00:00",
  "actor": "https://fed.brid.gy/snarfed.org",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Update",
  "object": {
    "updated": "2023-10-12T23:44:49.320885+00:00",
    "published": "2023-10-12T16:35:43-07:00",
    "content": "hi again to <a href=\"https://indieweb.social/users/snarfed\">@snarfed@indieweb.social</a> <a class=\"u-in-reply-to\" href=\"https://mas.to/@snarfed/109668310494520038\"> </a>",
    "url": "https://fed.brid.gy/r/https://snarfed.org/2023-10-12_snarfed-foo-bar-mas-to",
    "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-12_snarfed-foo-bar-mas-to",
    "type": "Note",
    "attributedTo": "https://fed.brid.gy/snarfed.org",
    "inReplyTo": "https://mas.to/users/snarfed/statuses/109668310494520038",
    "tag": [
      {
        "type": "Mention",
        "href": "https://indieweb.social/users/snarfed",
        "name": "@snarfed@indieweb.social"
      },
      {
        "type": "Mention",
        "href": "https://mas.to/users/snarfed"
      }
    ],
    "cc": [
      "https://indieweb.social/users/snarfed",
      "https://mas.to/users/snarfed",
      "https://www.w3.org/ns/activitystreams#Public",
      "https://mas.to/users/snarfed/followers"
    ],
    "to": [
      "https://www.w3.org/ns/activitystreams#Public"
    ]
  },
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "cc": [
    "https://indieweb.social/users/snarfed",
    "https://mas.to/users/snarfed",
    "https://www.w3.org/ns/activitystreams#Public",
    "https://mas.to/users/snarfed/followers"
  ],
}
```

**Comment by manton** [comment]
This is also a new post, not a reply to anything specific, so I don't include `inReplyTo`. Pretty similar for replies to a post, though. I think only `inReplyTo` is different.

**Comment by snarfed** [comment]
Woo, this is working! I honestly don't know what fixed it, but I won't argue. Docs are deploying now. Thank you @manton!

(Re your q's, I'd tried with `Create` too, and the empty second link was just a quirk of my site, unrelated to BF.)

**Comment by snarfed** [comment]
One silver lining of Mastodon's support here is that the AS2 `Mention` tag `href` isn't absolutely required to be the AP actor id, eg `https://indieweb.social/users/snarfed`. Mastodon also accepts the frontend URL that users see, eg `https://indieweb.social/@snarfed`.

**Comment by snarfed** [comment]
Reopening, this isn't working with Friendica groups, eg https://venera.social/profile/activitypubblueskybridge , which repost (`Announce`) your posts when you mention them. Sent this (trimmed here) to https://venera.social/inbox/activitypubblueskybridge a bit ago but it didn't work.

```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Create",
  "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-20_51173#bridgy-fed-create",
  "actor": "https://fed.brid.gy/snarfed.org",
  "published": "2023-10-20T19:27:57.093008+00:00",
  "object": {
    "type": "Note",
    "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-20_51173",
    "url": "https://fed.brid.gy/r/https://snarfed.org/2023-10-20_51173",
    "published": "2023-10-20T12:01:27-07:00",
    "content": "... cc <a href=\"https://venera.social/profile/activitypubblueskybridge\">@activitypubblueskybridge@venera.social</a>.</p>",
    "attributedTo": "https://fed.brid.gy/snarfed.org",
    "tag": [
      {
        "type": "Mention",
        "name": "@activitypubblueskybridge@venera.social",
        "href": "https://venera.social/profile/activitypubblueskybridge"
      }
    ],
    "cc": ["https://venera.social/profile/activitypubblueskybridge"],
    "to": ["https://www.w3.org/ns/activitystreams#Public"]
    "...",
  },
  "to": ["https://www.w3.org/ns/activitystreams#Public"]
}
```

**Comment by snarfed** [comment]
Here's the most recent post to that group, also trimmed a bit. I don't see an obvious difference yet. Notably, the `Mention` tags are identical.

You evidently have to follow the group before you can post to it, [which I did a while back](https://fed.brid.gy/web/snarfed.org/following), but maybe that follow didn't take?

```json
{
  "id": "https://indieweb.social/users/tchambers/statuses/111048163021099999",
  "type": "Note",
  "summary": null,
  "published": "2023-09-11T19:28:15Z",
  "url": "https://indieweb.social/@tchambers/111048163021099999",
  "attributedTo": "https://indieweb.social/users/tchambers",
  "to": ["https://www.w3.org/ns/activitystreams#Public"],
  "cc": [
    "https://indieweb.social/users/tchambers/followers",
    "https://venera.social/profile/activitypubblueskybridge",
    "https://venera.social/followers/activitypubblueskybridge",
    "https://fed.brid.gy/snarfed.org"
  ],
  "conversation": "tag:indieweb.social,2023-09-11:objectId=76434671:objectType=Conversation",
  "content": "<p>important work from <span class=\"h-card\" translate=\"no\"><a href=\"https://fed.brid.gy/r/https://snarfed.org/\" class=\"u-url mention\">@<span>snarfed.org</span></a></span> as part of his AP to BlueSky/AT bridge... comparing Nostr, AP, and AT Protocols.... Chime in if anything here needs tweaking or fixing...  cc: <span class=\"h-card\" translate=\"no\"><a href=\"https://venera.social/profile/activitypubblueskybridge\" class=\"u-url mention\">@<span>activitypubblueskybridge</span></a></span></p>",
  "tag": [
    {
      "type": "Mention",
      "href": "https://fed.brid.gy/snarfed.org",
      "name": "@snarfed.org@snarfed.org"
    },
    {
      "type": "Mention",
      "href": "https://venera.social/profile/activitypubblueskybridge",
      "name": "@activitypubblueskybridge@venera.social"
    }
  ],
  "..."
}
```

**Comment by snarfed** [comment]
Haven't found anything in their docs or GitHub searches so far:
* https://wiki.friendi.ca/docs/groups-and-privacy
* https://github.com/friendica/friendica/issues?q=group+mention+tag
* https://github.com/friendica/friendica/issues?q=groups+mention+activitypub

**Comment by snarfed** [comment]
Hmm, that group actually did end up reposting my mention post after all, just a while later, at 21:36:12 UTC, or at least I'm guessing based on this HTML in the repost:

```html
<span class="commented" style="display: none;">2023-10-22 02:16:53</span>
<span class="received" style="display: none;">2023-10-20 21:36:12</span>
<span class="created" style="display: none;">2023-10-20 19:01:27</span>
```

Not sure why the delay. But maybe this is working and I can close again after all.

**Comment by snarfed** [comment]
Following up here. Did some tests, turns out Mastodon's incoming mention handling is pretty relaxed.
* The `Mention` tag's `href` can be either the actor `id` (eg https://indieweb.social/users/snarfed ) or `url` (eg https://indieweb.social/@snarfed ).
* `name` isn't required at all.
* The mentioned actor doesn't need to be in `cc`.
* No HTML link in `content` is required.

So I'm relaxing BF's mention detection in #760.

**Comment by snarfed** [comment]
Can't believe I'm debugging this yet again. Trying to get DMs working for #880. The posts go through fine, and show as _🔒 Followers only_, but none of them show up in _Private mentions_. I've tried what feels like all possible combinations of mention tag href, name, and content. Ugh.

**Comment by snarfed** [comment]
Looked at this a bit. We need to convert the mention `href` to the mentioned actor's AP id. Right now we have eg ATProto record `at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3ksveg5r2qs24`:

```json
{
    "$type": "app.bsky.feed.post",
    "text": "maybe if @snarfed.indieweb.social.ap.brid.gy and #alf meet up",
    "createdAt": "2024-05-20T03:49:56.091Z",
    "facets": [{
      "$type": "app.bsky.richtext.facet",
      "features": [{
          "$type": "app.bsky.richtext.facet#mention",
          "did": "did:plc:bg5udl25mvzg3rt7l5n2hzet"
        }],
      "index": {
        "byteEnd": 44,
        "byteStart": 9
      }
    }],
  }
```

...to AS2:

```json
{
  "id": "https://bsky.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3ksveg5r2qs24",
  "url": "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3ksveg5r2qs24",
  "content": "<p>maybe if <a href=\"https://bsky.brid.gy/ap/https://bsky.app/profile/did:plc:bg5udl25mvzg3rt7l5n2hzet\">@snarfed.indieweb.social.ap.brid.gy</a> and <a href=\"https://bsky.brid.gy/hashtag/alf\">#alf</a> meet up</p>",
  "published": "2024-05-20T03:49:56.091Z",
  "attributedTo": "https://bsky.brid.gy/ap/did:plc:3ljmtyyjqcjee2kpewgsifvb",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Note",
  "tag": [{
      "type": "Mention",
      "name": "@snarfed.indieweb.social.ap.brid.gy",
      "href": "https://bsky.brid.gy/ap/https://bsky.app/profile/did:plc:bg5udl25mvzg3rt7l5n2hzet"
    }],
}
```
